SonarQube.Common/AnalysisConfig/AnalysisConfigExtensions.cs

```diff
@@ -43,80 +44,113 @@ public static string GetSetting(this AnalysisConfig config, string settingId, st
+        public static void SetConfigValue(this AnalysisConfig config, string settingId, string value)
+        {
+            SetValue(config, settingId, value);
+        }
+
+        /// <summary>
+        /// Returns a provider containing all of the analysis settings from the config.
+        /// Optionally includes settings downloaded from the SonarQube server.
+        /// </summary>
+        public static IAnalysisPropertyProvider GetAnalysisSettings(this AnalysisConfig config, bool includeServerSettings)
         {
             if (config == null)
             {
                 throw new ArgumentNullException("config");
             }
-            if (string.IsNullOrWhiteSpace(settingId))
-            {
-                throw new ArgumentNullException("settingId");
-            } 
 
-            result = null;
+            List<IAnalysisPropertyProvider> providers = new List<IAnalysisPropertyProvider>();
 
-            if (config.AdditionalSettings != null)
+            if (config.LocalSettings != null)
             {
-                result = config.AdditionalSettings.FirstOrDefault(ar => AnalysisSetting.SettingKeyComparer.Equals(settingId, ar.Id));
+                providers.Add(new ListPropertiesProvider(config.LocalSettings));
+            }
+            if (includeServerSettings && config.ServerSettings != null)
+            {
+                providers.Add(new ListPropertiesProvider(config.ServerSettings));
             }
-            return result != null;
-        }
 
-        /// <summary>
-        /// Sets the value of the additional setting. The setting will be added if it does not already exist.
-        /// </summary>
-        public static void SetInheritedValue(this AnalysisConfig config, string settingId, string value)
-        {
-            SetValue(config, settingId, value, true);
+            IAnalysisPropertyProvider provider = null;
+            switch(providers.Count)
+            {
+                case 0:
+                    provider = EmptyPropertyProvider.Instance;
+                    break;
+                case 1:
+                    provider = providers[0];
+                    break;
+                default:
+                    provider = new AggregatePropertiesProvider(providers.ToArray());
+                    break;
+            }
+
+            return provider;
         }
 
+        #endregion
+
+        #region Private methods
+
         /// <summary>
-        /// Sets the value of the additional setting. The setting will be added if it does not already exist.
+        /// Attempts to find and return the config setting with the specified id
         /// </summary>
-        public static void SetExplicitValue(this AnalysisConfig config, string settingId, string value)
+        /// <returns>True if the setting was found, otherwise false</returns>
+        private static bool TryGetSetting(this AnalysisConfig config, string settingId, out ConfigSetting result)
```

Comment:
This was a public method but is no longer required. I'll replace the arg checks with assertions and rename the method to TryGetConfigSetting.